Foundation/JSONSerialization.swift

```diff
@@ -28,6 +28,253 @@ extension JSONSerialization {
     }
 }
 
+fileprivate protocol _JSONValue {
+    func visit<Visitor: JSONVisitor>(with visitor: inout Visitor) throws
+}
+
+//@_specialize(where T == String)
+//extension<T: _JSONValue> [String : T] : _JSONValue {
+extension Dictionary : _JSONValue where Key == String, Value : _JSONValue {
+    
+//    @_specialize(where Visitor == JSONWriter)
+//    @_specialize(where Visitor == JSONValidator)
+    fileprivate func visit<Visitor: JSONVisitor>(with visitor: inout Visitor) throws {
+        try visitor.visit(self)
+    }
+}
+
+//@_specialize(where T == String)
+//extension<T: _JSONValue> [T] : _JSONValue {
+extension Array : _JSONValue where Element : _JSONValue {
+    
+//    @_specialize(where Visitor == JSONWriter)
+//    @_specialize(where Visitor == JSONValidator)
+    fileprivate func visit<Visitor: JSONVisitor>(with visitor: inout Visitor) throws {
+        try visitor.visit(self)
+    }
+}
+
+extension String : _JSONValue {
+    
+    @_specialize(where Visitor == JSONWriter)
```

Comment:
Does this generate a bunch of expensive type checks that outweigh the static dispatch? Who knows!